File: backend/HskLearn.Api/Models/Progress.cs
namespace HskLearn.Api.Models;

public record PracticeAttempt(
    string Type,
    string ItemId,
    double Score,
    Dictionary<string, object>? Details,
    DateTime Timestamp
);

public record ProgressSummary(
    int TotalWordsLearned,
    int TotalWords,
    TaskSummary Reading,
    TaskSummary Speaking,
    TaskSummary Writing,
    List<WeakWord> WeakWords
);

public record TaskSummary(
    int ItemsCompleted,
    double AverageScore
);

public record WeakWord(
    string Word,
    string Pinyin,
    double AverageScore
);

File: android/app/src/main/java/com/hsklearn/app/data/model/Progress.kt
package com.hsklearn.app.data.model

import kotlinx.serialization.Serializable

@Serializable
data class ProgressSummary(
    val totalWordsLearned: Int,
    val totalWords: Int,
    val reading: TaskSummary,
    val speaking: TaskSummary,
    val writing: TaskSummary,
    val weakWords: List<WeakWord>,
)

@Serializable
data class TaskSummary(
    val itemsCompleted: Int,
    val averageScore: Double,
)

@Serializable
data class WeakWord(
    val word: String,
    val pinyin: String,
    val averageScore: Double,
)

@Serializable
data class RecordProgressRequest(
    val type: String,
    val itemId: String,
    val score: Double,
)
